File: app-ios-nativa/Dominio/Pruebas/PruebasDominio/PruebasSincronizacion.swift
import Foundation
import Testing

@testable import Dominio

private func nuevo(
    _ url: String,
    titulo: String? = nil,
    etiquetas: [String] = [],
    en instante: MarcaDeTiempo = 100,
    generarId: GeneradorId
) -> Elemento {
    nuevoElementoLocal(
        DatosElementoNuevo(url: url, titulo: titulo, etiquetas: etiquetas),
        ahora: relojFijo(instante),
        generarId: generarId
    )
}

@Suite("Aplicar lo que baja del servidor")
struct PruebasAplicarPull {
    @Test("lo que no estaba se añade")
    func anadeLoNuevo() {
        let recibido = nuevo("https://a.com", generarId: generadorSecuencial())

        let cache = Sincronizacion.aplicarPull([:], recibidos: [recibido], pendientes: [:])

        #expect(cache[recibido.id] == recibido)
    }

    @Test("no pisa un cambio local que todavía está sin subir y es más nuevo")
    func respetaLoPendienteMasNuevo() {
        let base = nuevo("https://a.com", titulo: "Original", generarId: generadorSecuencial())
        let pendiente = base.conEtiquetas(["editado aquí"], ahora: relojFijo(300))
        let delServidor = base.conEtiquetas(["versión vieja del servidor"], ahora: relojFijo(200))

        let cache = Sincronizacion.aplicarPull(
            [base.id: base],
            recibidos: [delServidor],
            pendientes: [base.id: pendiente]
        )

        #expect(cache[base.id] == base)
    }

    @Test("si lo que baja es más nuevo que lo pendiente, sí se aplica")
    func aplicaLoMasNuevo() {
        let base = nuevo("https://a.com", titulo: "Original", generarId: generadorSecuencial())
        let pendiente = base.conEtiquetas(["editado aquí"], ahora: relojFijo(150))
        let delServidor = base.conEtiquetas(["ya venía de otro sitio"], ahora: relojFijo(500))

        let cache = Sincronizacion.aplicarPull(
            [base.id: base],
            recibidos: [delServidor],
            pendientes: [base.id: pendiente]
        )

        #expect(cache[base.id] == delServidor)
    }
}

@Suite("Aplicar la respuesta de lo que se sube")
struct PruebasAplicarPush {
    @Test("manda la versión del servidor, aunque sea más antigua")
    func mandaElServidor() {
        let local = nuevo("https://a.com", titulo: "Mío", generarId: generadorSecuencial())
        let definitivo = local.conEtiquetas(["el servidor ganó"], ahora: relojFijo(50))

        let cache = Sincronizacion.aplicarRespuestaPush(
            [local.id: local], definitivos: [definitivo])

        #expect(cache[local.id] == definitivo)
    }
}

@Suite("Qué se ve en la lista")
struct PruebasElementosVisibles {
    @Test("oculta los borrados y pone lo guardado más recientemente primero")
    func ordenYBorrados() {
        let generar = generadorSecuencial()
        let a = nuevo("https://a.com", en: 100, generarId: generar)
        let b = nuevo("https://b.com", en: 200, generarId: generar)
        let c = nuevo("https://c.com", en: 300, generarId: generar)
            .marcadoComoBorrado(ahora: relojFijo(400))

        let visibles = Sincronizacion.elementosVisibles([a.id: a, b.id: b, c.id: c])

        #expect(visibles.map(\.id) == [b.id, a.id])
    }

    @Test("retocar un enlace viejo no lo manda al principio de la lista")
    func retocarNoReordena() {
        let generar = generadorSecuencial()
        let viejo = nuevo("https://viejo.com", en: 100, generarId: generar)
        let nuevoEnlace = nuevo("https://nuevo.com", en: 200, generarId: generar)
        // Se le cambia una etiqueta al viejo: su fecha de modificación pasa a
        // ser la más alta de las dos, pero se guardó antes y ahí sigue.
        let viejoRetocado = viejo.conEtiquetas(["ocio"], ahora: relojFijo(900))

        let visibles = Sincronizacion.elementosVisibles([
            viejoRetocado.id: viejoRetocado, nuevoEnlace.id: nuevoEnlace,
        ])

        #expect(visibles.map(\.id) == [nuevoEnlace.id, viejoRetocado.id])
    }

    @Test("dos guardados en el mismo instante salen siempre en el mismo orden")
    func ordenEstable() {
        // Importar una biblioteca entera pone la misma fecha a todo. Sin
        // desempate, el diccionario los devolvería en cualquier orden y la
        // lista se recolocaría sola entre dos aperturas.
        let generar = generadorSecuencial()
        let a = nuevo("https://a.com", en: 100, generarId: generar)
        let b = nuevo("https://b.com", en: 100, generarId: generar)
        let cache = [a.id: a, b.id: b]

        let primero = Sincronizacion.elementosVisibles(cache).map(\.id)
        let segundo = Sincronizacion.elementosVisibles(cache).map(\.id)

        #expect(primero == segundo)
    }
}

@Suite("Etiquetas reservadas al sincronizar")
struct PruebasSincronizacionEtiquetas {
    @Test("no pisa una etiqueta local sin subir que es más nueva")
    func respetaLoPendiente() {
        let base = nuevaEtiquetaDefinida(
            nombre: "ocio",
            ahora: relojFijo(100),
            generarId: generadorSecuencial()
        )
        let pendiente = base.renombrada("tiempo libre", ahora: relojFijo(300))
        let delServidor = base.renombrada("ocio", ahora: relojFijo(200))

        let cache = Sincronizacion.aplicarPullEtiquetas(
            [base.id: base],
            recibidas: [delServidor],
            pendientes: [base.id: pendiente]
        )

        #expect(cache[base.id] == base)
    }

    @Test("las borradas no se ofrecen, y las demás salen por nombre")
    func visiblesOrdenadas() {
        let generar = generadorSecuencial()
        let ocio = nuevaEtiquetaDefinida(nombre: "ocio", generarId: generar)
        let casa = nuevaEtiquetaDefinida(nombre: "casa", generarId: generar)
        let ida = nuevaEtiquetaDefinida(nombre: "ida", generarId: generar)
            .marcadaComoBorrada()

        let visibles = Sincronizacion.etiquetasReservadasVisibles([
            ocio.id: ocio, casa.id: casa, ida.id: ida,
        ])

        #expect(visibles.map(\.nombre) == ["casa", "ocio"])
    }
}

@Suite("Cuándo toca sincronizar")
struct PruebasTocaSincronizar {
    @Test("sin ninguna sincronización previa, toca")
    func primeraVez() {
        // El cero no es «hace un instante»: es 1970.
        #expect(Sincronizacion.tocaSincronizar(ultima: 0, ahora: relojDelSistema()))
    }

    @Test("volver a la app dos veces seguidas no lanza dos sincronizaciones")
    func frenoEntreDos() {
        #expect(!Sincronizacion.tocaSincronizar(ultima: 1000, ahora: 2000, intervalo: 30_000))
    }

    @Test("pasado el intervalo vuelve a tocar")
    func pasadoElIntervalo() {
        #expect(Sincronizacion.tocaSincronizar(ultima: 1000, ahora: 31_000, intervalo: 30_000))
    }
}

File: app-android/dominio/src/test/kotlin/com/jmortizsilva/guardarenlaces/dominio/PruebasSincronizacion.kt
package com.jmortizsilva.guardarenlaces.dominio

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

private fun nuevo(
    url: String,
    titulo: String? = null,
    en: MarcaDeTiempo = 100,
    generarId: GeneradorId,
) =
    nuevoElementoLocal(
        DatosElementoNuevo(url = url, titulo = titulo),
        ahora = relojFijo(en),
        generarId = generarId,
    )

class PruebasAplicarPull {
    @Test
    fun `lo que no estaba se anade`() {
        val recibido = nuevo("https://a.com", generarId = generadorSecuencial())

        val cache = Sincronizacion.aplicarPull(emptyMap(), listOf(recibido), emptyMap())

        assertEquals(recibido, cache[recibido.id])
    }

    @Test
    fun `no pisa un cambio local que todavia esta sin subir y es mas nuevo`() {
        val base = nuevo("https://a.com", titulo = "Original", generarId = generadorSecuencial())
        val pendiente = base.conEtiquetas(listOf("editado aquí"), ahora = relojFijo(300))
        val delServidor = base.conEtiquetas(listOf("vieja del servidor"), ahora = relojFijo(200))

        val cache =
            Sincronizacion.aplicarPull(
                mapOf(base.id to base),
                listOf(delServidor),
                mapOf(base.id to pendiente),
            )

        assertEquals(base, cache[base.id])
    }

    @Test
    fun `si lo que baja es mas nuevo que lo pendiente, si se aplica`() {
        val base = nuevo("https://a.com", titulo = "Original", generarId = generadorSecuencial())
        val pendiente = base.conEtiquetas(listOf("editado aquí"), ahora = relojFijo(150))
        val delServidor = base.conEtiquetas(listOf("de otro sitio"), ahora = relojFijo(500))

        val cache =
            Sincronizacion.aplicarPull(
                mapOf(base.id to base),
                listOf(delServidor),
                mapOf(base.id to pendiente),
            )

        assertEquals(delServidor, cache[base.id])
    }
}

class PruebasAplicarPush {
    @Test
    fun `manda la version del servidor, aunque sea mas antigua`() {
        val local = nuevo("https://a.com", titulo = "Mío", generarId = generadorSecuencial())
        val definitivo = local.conEtiquetas(listOf("el servidor ganó"), ahora = relojFijo(50))

        val cache =
            Sincronizacion.aplicarRespuestaPush(mapOf(local.id to local), listOf(definitivo))

        assertEquals(definitivo, cache[local.id])
    }
}

class PruebasElementosVisibles {
    @Test
    fun `oculta los borrados y pone lo guardado mas recientemente primero`() {
        val generar = generadorSecuencial()
        val a = nuevo("https://a.com", en = 100, generarId = generar)
        val b = nuevo("https://b.com", en = 200, generarId = generar)
        val c =
            nuevo("https://c.com", en = 300, generarId = generar)
                .marcadoComoBorrado(ahora = relojFijo(400))

        val visibles = Sincronizacion.elementosVisibles(mapOf(a.id to a, b.id to b, c.id to c))

        assertEquals(listOf(b.id, a.id), visibles.map { it.id })
    }

    @Test
    fun `retocar un enlace viejo no lo manda al principio de la lista`() {
        val generar = generadorSecuencial()
        val viejo = nuevo("https://viejo.com", en = 100, generarId = generar)
        val nuevoEnlace = nuevo("https://nuevo.com", en = 200, generarId = generar)
        val viejoRetocado = viejo.conEtiquetas(listOf("ocio"), ahora = relojFijo(900))

        val visibles =
            Sincronizacion.elementosVisibles(
                mapOf(viejoRetocado.id to viejoRetocado, nuevoEnlace.id to nuevoEnlace)
            )

        assertEquals(listOf(nuevoEnlace.id, viejoRetocado.id), visibles.map { it.id })
    }

    @Test
    fun `dos guardados en el mismo instante salen por identificador`() {
        // Importar una biblioteca entera pone la misma fecha a todo. Sin desempate, la lista se
        // recolocaría sola entre dos aperturas.
        val b = Elemento(id = "b", url = "https://b.com", creadoEn = 100)
        val a = Elemento(id = "a", url = "https://a.com", creadoEn = 100)

        assertEquals(
            listOf("a", "b"),
            Sincronizacion.elementosVisibles(mapOf("b" to b, "a" to a)).map { it.id },
        )
    }
}

class PruebasSincronizacionEtiquetas {
    @Test
    fun `no pisa una etiqueta local sin subir que es mas nueva`() {
        val base =
            nuevaEtiquetaDefinida("ocio", ahora = relojFijo(100), generarId = generadorSecuencial())
        val pendiente = base.renombrada("tiempo libre", ahora = relojFijo(300))
        val delServidor = base.renombrada("ocio", ahora = relojFijo(200))

        val cache =
            Sincronizacion.aplicarPullEtiquetas(
                mapOf(base.id to base),
                listOf(delServidor),
                mapOf(base.id to pendiente),
            )

        assertEquals(base, cache[base.id])
    }

    @Test
    fun `las borradas no se ofrecen, y las demas salen por nombre`() {
        val generar = generadorSecuencial()
        val ocio = nuevaEtiquetaDefinida("ocio", generarId = generar)
        val casa = nuevaEtiquetaDefinida("casa", generarId = generar)
        val ida = nuevaEtiquetaDefinida("ida", generarId = generar).marcadaComoBorrada()

        val visibles =
            Sincronizacion.etiquetasReservadasVisibles(
                mapOf(ocio.id to ocio, casa.id to casa, ida.id to ida)
            )

        assertEquals(listOf("casa", "ocio"), visibles.map { it.nombre })
    }

    @Test
    fun `la respuesta del servidor manda tambien en las etiquetas`() {
        val local =
            nuevaEtiquetaDefinida("ocio", ahora = relojFijo(300), generarId = generadorSecuencial())
        val definitiva = local.renombrada("ganó el servidor", ahora = relojFijo(100))

        val cache =
            Sincronizacion.aplicarRespuestaPushEtiquetas(
                mapOf(local.id to local),
                listOf(definitiva),
            )

        assertEquals(definitiva, cache[local.id])
    }
}

class PruebasTocaSincronizar {
    @Test
    fun `sin ninguna sincronizacion previa, toca`() {
        // El cero no es «hace un instante»: es 1970.
        assertTrue(Sincronizacion.tocaSincronizar(ultima = 0, ahora = relojDelSistema()))
    }

    @Test
    fun `volver a la app dos veces seguidas no lanza dos sincronizaciones`() {
        assertFalse(Sincronizacion.tocaSincronizar(ultima = 1000, ahora = 2000, intervalo = 30_000))
    }

    @Test
    fun `pasado el intervalo vuelve a tocar`() {
        assertTrue(
            Sincronizacion.tocaSincronizar(ultima = 1000, ahora = 31_000, intervalo = 30_000)
        )
    }
}
